src/graph.rs

```diff
@@ -460,6 +467,153 @@ where
+            let target = *edge.target();
+
+            let source_result = addresses.iter().position(|&r| r == source);
+            if source_result.is_none() {
+                continue;
+            }
+
+            let target_result = addresses.iter().position(|&r| r == target);
+            if target_result.is_none() {
+                continue;
+            }
+
+            let source_index = source_result.unwrap();
+            let target_index = target_result.unwrap();
+            agraph[source_index].push(target_index);
+            agraph[target_index].push(source_index);
+        }
+        agraph
+    }
+
+    // This method returns the closeness and betweenness for a given AGraph.
+    //
+    // Closeness:
+    //   for each node
+    //     for all other nodes
+    //       find all shortest paths to that other node
+    //         accumlate all path lens
+    //         accumulate number of paths.
+    //   compute average path length
+    //
+    //
+    //  Betweenness:
+    //    When a shortest path is found
+    //      for all nodes in-between (i.e., not an end point)
+    //        increment their betweenness value
+    //
+    pub fn compute_betweenness_and_closeness(&self, agraph: &AGraph) -> (Vec<u32>, Vec<f64>) {
+        let num_nodes = agraph.len();
+
+        let mut betweenness: Vec<u32> = vec![0; num_nodes];
+        let mut closeness: Vec<f64> = vec![0.0; num_nodes];
+        let mut total_path_length: Vec<u32> = vec![0; num_nodes];
+        let mut num_paths: Vec<u32> = vec![0; num_nodes];
+
+        for i in 0..num_nodes - 1 {
+            let mut visited: Vec<bool> = vec![false; num_nodes];
+            let mut found_or_not: Vec<bool> = vec![false; num_nodes];
+            let mut search_list: Vec<usize> = Vec::new();
+
+            // mark node i and all those before i as visited
+            for j in 0..i + 1 {
+                found_or_not[j] = true;
+            }
+            for j in i + 1..num_nodes {
+                search_list.push(j);
+                found_or_not[j] = false;
+            }
+
+            while search_list.len() > 0 {
+                // 0. OUR MAIN SEARCH LOOP:  I and J
+                // 1. we search for path between i and j.  We're done when we find j
+                // 2. any short paths we find along the way, they get handled and removed from search list
+                // 3. along the way, we appropriately mark any between nodes
+                // 4. we also determine if no path exists (disconnected graph case)
+                let mut done = false;
+                let j = search_list[0];
+                for x in 0..num_nodes {
+                    visited[x] = x == i;
+                }
+                let mut pathlen: u32 = 1;
+                let mut queue_list = Vec::new();
+                queue_list.push(i);
+
+                while !done {
+                    let mut this_round_found: Vec<usize> = Vec::new();
+                    let mut queue_me = Vec::new();
```

Comment:
These two collections will either need comments specifying what they are or clearer naming as what they're for isn't immediately obvious without reading the implementation details.

Reply:
done